### ISSUE TITLE
Could you share Cifar100 pretrain module to us?

### ISSUE THREAD

**Issue by FrancisYizhang** [titu1994/DenseNet#20]
Dear @titu1994 

  I want to train a baseline module on cifar100, dense-40-12. But, I can not obtain performance like the paper. Could you share your hyper-parameters or pre-train module on cifar100?

  Thanks a milion.

**Comment by titu1994** [comment]
Try matching the hyper parameters of the paper as much as you can, along with preprocessing and image augmentation techniques. Keras does not offer image crops or other such augmentation so it is not possible to match it exactly. 

I do not have weights for Cifar 100. You might want to look into other Keras repos for weights.

**Comment by ahundt** [comment]
The authors now [suggest doing a wide version](https://github.com/liuzhuang13/DenseNet) to improve what's in the paper. Image crops can be done with tf tensors that implement it before feeding it to your model. see [mnist_tfrecord.py](https://github.com/fchollet/keras/blob/master/examples/mnist_tfrecord.py) for an example.

**Comment by titu1994** [comment]
Oh so the DenseNet BC architecture with greater width? That's certainly interesting, but my GPU can barely handle 40-12 currently. I am wondering how to incorporate the shared batch norm memory efficient implementation from the torch repo, but I can't make any headway. 

Also class assignments are starting to back up so can't pay much attention to this either, sadly. Having a memory efficient implementation of DenseNet for Keras is absolutely paramount in my opinion, as it would allow for deep dense networks. 

But until I find some way to incorporate the wierd shared batchnorm from the torch codebase, that's a pipe dream.

**Comment by DerekGrant** [comment]
I test cifar 100, it reach 60% after 200 epochs with depth 40, is there any parameter need to modify expect depth?

**Comment by titu1994** [comment]
That's wierd. It should reach the high 70s with that many epochs and 40-12 model. How did you normalize the data ? Also, did you use learning rate schedule to reduce the learning rate ?

**Comment by titu1994** [comment]
I'm going to review the latest DenseNet paper and reimplenent this based on the PyTorch version. 

Expect breaking changes in the coming days. Once I finish, I will try to train on Cifar 10 and 100. If the performance is similar to the paper, I will update the weights and the models. 

But since it is going to be based on the PyTorch version, I may need to add a custom scaling layer and so on. A lot of breaking changes.

**Comment by titu1994** [comment]
Most probably, scale layers wont be needed. Keras BatchNorm has scaling inbuilt. I will have to port the weights though, which will take time.

**Comment by ahundt** [comment]
@titu1994 will you be updating it in keras-contrib too? That's the one I actually use

**Comment by titu1994** [comment]
If I can get it to work, yes. I was able to exactly duplicate the code to the PyTorch version, matching both the Cifar and ImageNet versions exactly.

However I have to use my GPU for other research projects so I will not be able to use it for training Cifar anytime soon. 

I will probably port the ImageNet weights and post that to Contrib.

**Comment by ahundt** [comment]
cool, I had been making some progress on that at https://github.com/ahundt/keras-contrib/releases and https://github.com/ahundt/keras-contrib/tree/cnn_finetune, but I had to put it on the back burner and there are some small model differences, what you're doing might be a better approach.

If you could, make it easy to set all the configuration parameters correctly for the pretrained model to be loaded.

**Comment by titu1994** [comment]
@ahundt Just finished porting all the weights from https://github.com/flyyufelix/DenseNet-Keras to the repo. @flyyufelix, great work on porting the Caffe weights to Keras.

I was able to remove the redundant Scale Layer, since Keras Batchnorm has that inbuilt. All 3 weight files were ported over and can do training/inference correctly now.

I've also added pre-set models, such as DenseNetImageNet121, 161, 169 (which have weights), and 201 and 269 which dont have weights yet (I am on windows, cannot port caffe weights).

There is now a new flag - subsample_initial_block, which decides whether the first convolution should have stride 2 or not and be followed by BN-Relu-MaxPool2D or not. This is necessary as CIFAR DenseNets do not have the BN-Relu-MaxPool2D and have a stride 1 initial convolution (otherwise the spatial resolution is 8x8 right off the bat for the 1st dense block, which is too small to learn properly, though I have not tested it). This flag default to False, which means the default DenseNet model defaults to CIFAR mode. 

A very large number of changes were made to the repository, perhaps quite a few breaking changes. The older CIFAR 10 weights will no longer work. That will have to be substituted by the new ImageNet models. If and when someone manages to train on CIFAR, I will update the codebase to accommodate them as well.

Finally, porting this to Keras Contrib is a huge job. It will take time to do it correctly. I suggest starting off using it from the repo for now, and later on the integration with contrib should be seamless @ahundt

**Comment by ahundt** [comment]
looks cool, thanks! 

What's the barrier for keras-contrib, mostly formatting or perhaps some of the added options? (just for clarity)

**Comment by titu1994** [comment]
Majority is the code cleanup - the new parameters, the pre-built models can be copy pasted, the imports in contrib are different. Minor things, but coupled with my limited time schedule, it is gonna take time.

Mainly, its code cleanup which will take perhaps the longest. It's sadly not possible to copy paste the entire script in my repo to the contrib script.
